man/pick.man
.TH PICK %manext1% 2016-03-12 "%nmhversion%"
.
.\" %nmhwarning%
.
.SH NAME
pick \- search nmh messages
.SH SYNOPSIS
.HP 5
.na
.B pick
.RB [ \-help ]
.RB [ \-version ]
.RI [ +folder ]
.RI [ msgs ]
.RB [ \-reverse
\&...]
.RB [ \-and
\&...]
.RB [ \-or
\&...]
.RB [ \-not
\&...]
.RB [ \-lbrace
\&...
.BR \-rbrace ]
.RB [ \-\|\-component
.IR pattern ]
.RB [ \-cc
.IR pattern ]
.RB [ \-date
.IR pattern ]
.RB [ \-from
.IR pattern ]
.RB [ \-search
.IR pattern ]
.RB [ \-subject
.IR pattern ]
.RB [ \-to
.IR pattern ]
.RB [ \-after
.IR date ]
.RB [ \-before
.IR date ]
.RB [ \-datefield
.IR field ]
.RB [ \-sequence
.I name
\&...]
.RB [ \-nosequence ]
.RB [ \-public " | " \-nopublic ]
.RB [ \-zero " | " \-nozero ]
.RB [ \-list " | " \-nolist ]
.RB [ \-debug ]
.PP
typical usage:
.PP
.RS 5
.nf
scan\0\`pick\0\-from\0jones\`
pick\0\-to\0holloway\0\-sequence\0select
show\0\`pick\0\-before\0friday\`
.fi
.RE
.ad
.SH DESCRIPTION
.B pick
searches within a folder for messages with the specified
contents, and then identifies those messages.  Two types of search
primitives are available: pattern matching and date constraint
operations.
.PP
A modified
.IR grep (1)
is used to perform the matching, so the full regular expression (see
.IR ed (1))
facility is available within
.IR pattern .
With
.BR \-search ,
.I pattern
is used directly, and with the others, the grep pattern constructed is:
.PP
.RS 5
`component[ \\t]*:.*pattern'
.RE
.PP
This means that the pattern specified for a
.B \-search
will be found everywhere in the message, including the header and the body,
while the other pattern matching requests are limited to the single specified
component.  The expression
.PP
.RS 5
`\-\|\-component\ pattern'
.RE
.PP
is a shorthand for specifying
.PP
.RS 5
`\-search \*(lqcomponent[ \\t]*:.*pattern\*(rq\ '
.RE
.PP
It is used to pick a component which is not one of \*(lqTo:\*(rq,
\*(lqcc:\*(rq, \*(lqDate:\*(rq, \*(lqFrom:\*(rq, or \*(lqSubject:\*(rq.
An example is
.RB \*(lq "pick\0\-\|\-reply\-to\0pooh" \*(rq.
.PP
Pattern matching is performed on a per-line basis.  Within the header
of the message, each component is treated as one long line, but in the
body, each line is separate.  Lower-case letters in the search pattern
will match either lower or upper case in the message, while upper case
will match only upper case.
.PP
Note that since the
.B \-date
switch is a pattern matching operation (as described above),
to find messages sent on a certain date the pattern string must match
the text of the \*(lqDate:\*(rq field of the message.
.PP
Independent of any pattern matching operations requested, the switches
.B \-after
.I date
or
.B \-before
.I date
may also be used to introduce date/time constraints on all of the messages.
By default, the \*(lqDate:\*(rq field is consulted, but if another
date-yielding field (such as \*(lqBB\-Posted:\*(rq or
\*(lqDelivery\-Date:\*(rq) should be used, the
.B \-datefield
.I field
switch may be used.
.PP
With
.B \-before
and
.BR \-after ,
.B pick
will actually parse the date fields in each of the messages specified in
`msgs' and compare them to the date/time specified.  If
.B \-after
is given, then only those messages whose \*(lqDate:\*(rq field value is
chronologically after the date specified will be considered.  The
.B \-before
switch specifies the complementary action.
.PP
Both the
.B \-after
and
.B \-before
switches take legal RFC 822-style date specifications as arguments.
.B pick
will default certain missing fields so that the entire date need not
be specified.  These fields are (in order of defaulting): timezone,
time and timezone, date, date and timezone.
All defaults are taken from the current date, time, and timezone.
.PP
In addition to RFC 822-style dates,
.B pick
will also recognize any of the days of the week (\*(lqsunday\*(rq,
\*(lqmonday\*(rq, and so on), and the special dates \*(lqtoday\*(rq,
\*(lqyesterday\*(rq (24 hours ago), and \*(lqtomorrow\*(rq
(24 hours from now).
All days of the week are judged to refer to a day in the past
(e.g., telling \fIpick\fR \*(lqsaturday\*(rq on a \*(lqtuesday\*(rq
means \*(lqlast\ saturday\*(rq not \*(lqthis\ saturday\*(rq).
.PP
Finally, in addition to these special specifications,
.B pick
will also honor a specification of the form \*(lq\-dd\*(rq, which means
\*(lqdd days ago\*(rq.
.PP
Use the
.B \-reverse
switch to make
.B pick
find matching messages in reverse order, working from the highest message
number down to the lowest.  This can be useful in searching for recent
messages in large folders, for example,
.PP
.RS 5
.nf
pick\0\-reverse\0\-from\0frated\0|\0xargs\0\-n1\0scan
.fi
.RE
.PP
.B pick
supports complex boolean operations on the searching primitives with the
.BR \-and ,
.BR \-or ,
.BR \-not ,
and
.B \-lbrace
.B \&...
.B \-rbrace
switches.
For example,
.PP
.RS 5
.nf
pick\0\-after\0yesterday\0\-and
     \-lbrace\0\-from\0freida\0\-or\0\-from\0fear\0\-rbrace
.fi
.RE
.PP
identifies messages recently sent by \*(lqfrieda\*(rq or \*(lqfear\*(rq.
.PP
The matching primitives take precedence over the
.B \-not
switch, which in turn takes precedence over
.B \-and
which in turn takes precedence over
.BR \-or .
To override the default precedence, the
.B \-lbrace
and
.B \-rbrace
switches are provided, which act just like opening and closing
parentheses in logical expressions.
.PP
If no search criteria are given, all the messages specified on the
command line are selected (this defaults to \*(lqall\*(rq).
.PP
Once the search has been performed, if the
.B \-list
switch is given, the message numbers of the selected messages are
written to the standard output separated by newlines.  This is
.I extremely
useful for quickly generating arguments for other
.B nmh
programs by using the \*(lqbackquoting\*(rq syntax of the shell.
For example, the command
.PP
.RS 5
scan\0\`pick\0+todo\0\-after\0\*(lq31 Mar 83 0123 PST\*(rq\`
.RE
.PP
says to
.B scan
those messages in the indicated folder which meet the appropriate
criterion.  Note that since
.BR pick 's
context changes are written out prior to
.BR scan 's
invocation, you need not give the folder argument to
.B scan
as well.
.PP
The
.B \-sequence
.I name
switch may be given once for each sequence the user wishes to define.
For each sequence named, that sequence will be defined to mean exactly
those messages selected by
.BR pick .
For example,
.PP
.RS 5
pick\0\-from\0frated\0\-seq\0fred
.RE
.PP
defines a new message sequence for the current folder called
\*(lqfred\*(rq which contains exactly those messages that were selected.
.PP
The
.B \-nosequence
switch will disable all previously named sequences, allowing
those established by a profile component to be overridden.
.PP
By default,
.B pick
will zero a sequence before adding it.  This action can be disabled with the
.B \-nozero
switch, which means that the messages selected by
.B pick
will be added to the sequence, if it already exists, and any messages
already a part of that sequence will remain so.
.PP
The
.B \-public
and
.B \-nopublic
switches are used by
.B pick
in the same way
.B mark
uses them.
.PP
The
.B \-debug
switch causes pick to output a representation of the search pattern.
.SS "Output when no messages are matched"
If
.B pick
is used in a backquoted operation, such as
.PP
.RS 5
scan\0\`pick\0\-from\0jones\`
.RE
.PP
and
.B pick
selects no messages (e.g., no messages are from \*(lqjones\*(rq),
then the shell will still run the outer command (e.g.,
.BR scan ).
Since no messages were matched,
.B pick
produced no output, and the argument given to the outer command as a
result of backquoting
.B pick
is empty.  In the case of
.B nmh
programs, the outer command now acts as if the default `msg' or `msgs'
should be used (e.g., \*(lqall\*(rq in the case of
.BR scan ).
To prevent this unexpected behavior, if
.B \-list
was given, and if its standard output is not a tty, then
.B pick
outputs the illegal message number \*(lq0\*(rq when it fails.
This lets the outer command fail gracefully as well.
.SH FILES
.fc ^ ~
.nf
.ta \w'%nmhetcdir%/ExtraBigFileName  'u
^$HOME/.mh_profile~^The user profile
.fi
.SH "PROFILE COMPONENTS"
.fc ^ ~
.nf
.ta 2.4i
.ta \w'ExtraBigProfileName  'u
^Path:~^To determine the user's nmh directory
^Current\-Folder:~^To find the default current folder
.fi
.SH "SEE ALSO"
.IR mark (1)
.SH DEFAULTS
.nf
.RB ` +folder "' defaults to the current folder"
.RB ` msgs "' defaults to all"
.RB ` "\-datefield date" '
.RB ` \-zero '
.RB ` \-list "' is the default if no `\-sequence', `\-nolist' otherwise"
.fi
.SH CONTEXT
If a folder is given, it will become the current folder.
.SH HISTORY
In previous versions of
.BR MH ,
the
.B pick
command would
.BR show ,
.BR scan ,
or
.B refile
the selected messages.  This was rather
\*(lqinverted logic\*(rq from the Unix point of view, so
.B pick
was changed to define sequences and output those sequences.  Hence,
.B pick
can be used to generate the arguments for all other
.B MH
commands, instead of giving
.B pick
endless switches for invoking those commands
itself.
.PP
Also, previous versions of
.B pick
balked if you didn't specify
a search string or a date/time constraint.  The current version does
not, and merely matches the messages you specify.  This lets you type
something like:
.PP
.RS 5
show\0\`pick\0last:20\0\-seq\0fear\`
.RE
.PP
instead of typing
.PP
.RS 5
.nf
mark\0\-add\0\-nozero\0\-seq\0fear\0last:20
show\0fear
.fi
.RE
.PP
Finally, timezones used to be ignored when comparing dates: they aren't
any more.
.SH "HELPFUL HINTS"
Use
.RB \*(lq "pick sequence \-list" \*(rq
to enumerate the messages in a sequence
(such as for use by a shell script).
.SH BUGS
Any occurrence of
.B \-datefield
must occur prior to the
.B \-after
or
.B \-before
switch it applies to.
.PP
The pattern syntax \*(lq[l-r]\*(rq is not supported; each letter to be
matched must be included within the square brackets.


NEWS
# NEWS - Release notes for nmh 1.7
#

Welcome to nmh, the new version of the classic MH mail handling system.
It's been over three years since the last release of nmh, and there have
been a number of significant changes since the last release.  Long-time
MH and nmh uses should read careful the NOTEABLE CHANGES section, as there
are some significant changes to nmh behavior.  Otherwise, please see the
README and INSTALL files for help on getting started with nmh.

---------------
NOTABLE CHANGES
---------------

The largest notable changes in the 1.7 release are:

- Complete unification of network security support.  All network protocols
  (currently, POP and SMTP) have been refactored to use a common set of
  security routines.  This means all protocols support all SASL mechanisms
  (via the Cyrus-SASL library) and TLS.  TLS support has been strengthened
  to perform certificate name validation and to require TLS 1.1 as a
  minimum protocol.  Also, all protocols can make use of the OAuth2/XOAUTH
  SASL mechanism, which is supported by gmail.
- send(1) now supports adding switches to post(8) based on the address or
  domain of the email address in the From: header; this more easily allows
  users to support multiple identities.
- A generic facility for passing arguments to filter programs in repl(1)
  by use of the -convertargs switch.
- Native support for the manipulation of iCalendar requests; see mhical(1)
  for more details.

------------
NEW FEATURES
------------

The following are new features for the 1.7 release of nmh:

- When building from source, configure will derive ${prefix} from an existing
  nmh installation if it finds one in your $PATH.
- Added welcome message when nmh detects that its version changed.
- The default locations for configuration files and support binaries
  have been changed.  Configuration files now install into ${sysconfdir}/nmh,
  and support binaries are placed in ${libexecdir}/nmh.  If you are upgrading
  an existing installation you should look for old configuration files in
  ${sysconfdir} and merge any local customizations into the new files in
  ${sysconfdir}/nmh, then remove the old files.  ${libdir} will also contain
  obsolete support programs that should be removed.
- All TLS connections now perform certificate validation (including hostname
  matching) by default; can be disabled on a per-application basis.
- post now defaults to port 587 on 'smtp' message submission.
- A value of 0 for the width switch of scan(1), inc(1), ap(1), dp(1),
  fmttest(1), and mhl(1) now means as many characters as the format
  engine can produce [Bug #15274].  That amount is limited by internal
  buffers.
- If a component has trailing whitespace, e.g., body:component="> ",
  mhl now trims that whitespace off when filtering blank text lines.
- An "rtrim" flag has been added to mhl to remove any trailing
  whitespace from filtered text lines.  A corresponding "nortrim" flag
  has also been added.
- Added getmymbox and getmyaddr mh-format(5) function escapes.
- New -[no]changecur, -fixtype, -decodetypes, and -[no]crlflinebreaks switches
  have been added to mhfixmsg(1).
- mhfixmsg now removes an extraneous trailing semicolon from header
  parameter lists.
- Added -convertargs switch to repl(1), to pass arguments to programs
  specified in the user's profile or mhn.defaults to convert message
  content.
- Added mhical(1), to display, reply to, and cancel iCalendar (RFC 5545)
  event requests.
- Added multiply mh-format(5) function.
- "mhparam bindir" prints the path to the directory containing the public
  executables (${bindir}).
- New "-prefer" switch for mhshow (and mhlist and mhshow), to allow specifying
  the preferred content types to show, if present in a multipart alternative.
- mh-format now has %(kilo) and %(kibi) functions, to allow printing
  numbers with SI or IEC quantities, e.g. "10K", "2.3Mi".
- Support for the -sendmail flag to send/post to change the sendmail
  binary when using the sendmail/pipe MTS.
- Added support to send(1) to specify switches to post(1) based on address or
  domain name in From: header line in message draft.
- post(8) -snoop now attempts to decode base64-encoded SMTP traffic.
- folder(1) -nocreate now prints a warning message for a non-existent folder.
- mhfixmsg(1) now allows -decodetext binary, though 8bit is still the default.
- inc(1) and msgchk(1) now support TLS encryption natively.
- All network protocols support the XOAUTH authentication mechanism.
- Support for SMTPUTF8 (RFC 6531) has been added.  mhshow(1) already supported
  RFC 6532, assuming all 8-bit message header field bodies are UTF-8 and use
  of a UTF-8 locale.
- mhfixmsg now replaces RFC 2047 encoding with RFC 2231 encoding of name and
  filename parameters in Content-Type and Content-Disposition headers,
  respectively.
- If a message body contains 8-bit bytes, post(8) uses SMTP 8BITMIME if the
  server supports it.  If not, post fails with a message to the user to
  encode the message for 7-bit transport.
- Fewer lseek(2)s will be used when reading headers in the common case.
- ./configure's --enable-debug has been removed;  it did nothing.
- configure now defaults to enabling each of TLS and Cyrus SASL if the
  necessary headers and libraries are found.
- Moved build_nmh to top-level directory.
- Better error reporting for connections to network services.

---------
BUG FIXES
---------

- The format scanner no longer subtracts 1 from the width.  This has the
  effect of no longer counting the trailing newline in the output of
  scan(1), inc(1), and the other programs that rely on the format scanner.
- The first character of some very short (less than 4 characters) message
  bodies is no longer dropped.
- Single-character headers can be reliably formatted, etc., instead of
  apparently being missing.
- mhfixmsg now adds a Content-Transfer-Encoding header at the message level,
  if needed after decoding text parts.

uip/pick.c
/* pick.c -- search for messages by content
 *
 * This code is Copyright (c) 2002, 2008, by the authors of nmh.  See the
 * COPYRIGHT file in the root directory of the nmh distribution for
 * complete copyright information.
 */

#include <h/mh.h>
#include <h/tws.h>
#include <h/picksbr.h>
#include "h/done.h"
#include <h/utils.h>
#include "sbr/m_maildir.h"

#define PICK_SWITCHES \
    X("reverse", 0, REVSW) \
    X("and", 0, ANDSW) \
    X("or", 0, ORSW) \
    X("not", 0, NOTSW) \
    X("lbrace", 0, LBRSW) \
    X("rbrace", 0, RBRSW) \
    X("cc  pattern", 0, CCSW) \
    X("date  pattern", 0, DATESW) \
    X("from  pattern", 0, FROMSW) \
    X("search  pattern", 0, SRCHSW) \
    X("subject  pattern", 0, SUBJSW) \
    X("to  pattern", 0, TOSW) \
    X("-othercomponent  pattern", 0, OTHRSW) \
    X("after date", 0, AFTRSW) \
    X("before date", 0, BEFRSW) \
    X("datefield field", 5, DATFDSW) \
    X("sequence name", 0, SEQSW) \
    X("nosequence", 0, NSEQSW) \
    X("public", 0, PUBLSW) \
    X("nopublic", 0, NPUBLSW) \
    X("zero", 0, ZEROSW) \
    X("nozero", 0, NZEROSW) \
    X("list", 0, LISTSW) \
    X("nolist", 0, NLISTSW) \
    X("debug", 0, DEBUGSW) \
    X("version", 0, VERSIONSW) \
    X("help", 0, HELPSW) \

#define X(sw, minchars, id) id,
DEFINE_SWITCH_ENUM(PICK);
#undef X

#define X(sw, minchars, id) { sw, minchars, id },
DEFINE_SWITCH_ARRAY(PICK, switches);
#undef X

static int listsw = -1;

static void putzero_done (int) NORETURN;

int
main (int argc, char **argv)
{
    int publicsw = -1, zerosw = 1, vecp = 0;
    size_t seqp = 0;
    int msgnum;
    char *maildir, *folder = NULL, buf[100];
    char *cp, **argp, **arguments;
    svector_t seqs = svector_create (0);
    char *vec[MAXARGS];
    struct msgs_array msgs = { 0, 0, NULL };
    struct msgnum_array nums = { 0, 0, NULL };
    struct msgs *mp, *mp2;
    FILE *fp;
    int debug = 0;
    int reverse = 0;
    int start, end, inc;

    if (nmh_init(argv[0], 1)) { return 1; }

    set_done(putzero_done);

    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

    while ((cp = *argp++)) {
	if (*cp == '-') {
	    if (*++cp == '-') {
		vec[vecp++] = --cp;
		goto pattern;
	    }
	    switch (smatch (cp, switches)) {
	    case AMBIGSW: 
		ambigsw (cp, switches);
		listsw = 0;	/* HACK */
		done (1);
	    case UNKWNSW: 
		adios (NULL, "-%s unknown", cp);

	    case HELPSW: 
		snprintf (buf, sizeof(buf), "%s [+folder] [msgs] [switches]",
			  invo_name);
		print_help (buf, switches, 1);
		listsw = 0;	/* HACK */
		done (0);
	    case VERSIONSW:
		print_version(invo_name);
		listsw = 0;	/* HACK */
		done (0);

            case REVSW:
                reverse = 1;
                continue;

	    case CCSW: 
	    case DATESW: 
	    case FROMSW: 
	    case SUBJSW: 
	    case TOSW: 
	    case DATFDSW: 
	    case AFTRSW: 
	    case BEFRSW: 
	    case SRCHSW: 
		vec[vecp++] = --cp;
	    pattern:
		if (!(cp = *argp++))/* allow -xyz arguments */
		    adios (NULL, "missing argument to %s", argp[-2]);
		vec[vecp++] = cp;
		continue;
	    case OTHRSW: 
		adios (NULL, "internal error!");

	    case ANDSW:
	    case ORSW:
	    case NOTSW:
	    case LBRSW:
	    case RBRSW:
		vec[vecp++] = --cp;
		continue;

	    case SEQSW: 
		if (!(cp = *argp++) || *cp == '-')
		    adios (NULL, "missing argument to %s", argp[-2]);

                if (!seq_nameok (cp))
                  done (1);

		svector_push_back (seqs, cp);
		seqp++;
		continue;
	    case NSEQSW:
		seqp = 0;
		continue;
	    case PUBLSW: 
		publicsw = 1;
		continue;
	    case NPUBLSW: 
		publicsw = 0;
		continue;
	    case ZEROSW: 
		zerosw++;
		continue;
	    case NZEROSW: 
		zerosw = 0;
		continue;

	    case LISTSW: 
		listsw = 1;
		continue;
	    case NLISTSW: 
		listsw = 0;
		continue;

	    case DEBUGSW:
		++debug;
		continue;
	    }
	}
	if (*cp == '+' || *cp == '@') {
	    if (folder)
		adios (NULL, "only one folder at a time!");
            folder = pluspath (cp);
	} else
		app_msgarg(&msgs, cp);
    }
    vec[vecp] = NULL;

    if (!context_find ("path"))
	free (path ("./", TFOLDER));

    /*
     * If we didn't specify which messages to search,
     * then search the whole folder.
     */
    if (!msgs.size)
	app_msgarg(&msgs, "all");

    if (!folder)
	folder = getfolder (1);
    maildir = m_maildir (folder);

    if (chdir (maildir) == NOTOK)
	adios (maildir, "unable to change directory to");

    /* read folder and create message structure */
    if (!(mp = folder_read (folder, 0)))
	adios (NULL, "unable to read folder %s", folder);

    /* check for empty folder */
    if (mp->nummsg == 0)
	adios (NULL, "no messages in %s", folder);

    /* parse all the message ranges/sequences and set SELECTED */
    for (msgnum = 0; msgnum < msgs.size; msgnum++)
	if (!m_convert (mp, msgs.msgs[msgnum]))
	    done (1);

    /*
     * If we aren't saving the results to a sequence,
     * we default to list the results.
     */
    if (listsw == -1)
	listsw = !seqp;

    if (publicsw == 1 && is_readonly(mp))
	adios (NULL, "folder %s is read-only, so -public not allowed", folder);

    if (!pcompile (vec, NULL))
	done (1);

    /* If printing message numbers to standard out, force line buffering on.
     */
    if (listsw)
	setvbuf (stdout, NULL, _IOLBF, 0);

    /*
     * Scan through all the SELECTED messages and check for a
     * match.  If there is NOT a match, then add it to a list to
     * remove from the final sequence (it will make sense later)
     */
    if (!reverse) { /* Overflow or underflow is fine. */
        start = mp->lowsel;
        end = mp->hghsel + 1;
        inc = 1;
    } else {
        start = mp->hghsel;
        end = mp->lowsel - 1;
        inc = -1;
    }
    for (msgnum = start; msgnum != end; msgnum += inc) {
	if (is_selected (mp, msgnum)) {
	    if ((fp = fopen (cp = m_name (msgnum), "r")) == NULL)
		admonish (cp, "unable to read message");
	    if (fp && pmatches (fp, msgnum, 0L, 0L, debug)) {
		if (listsw)
		    puts(m_name (msgnum));
	    } else {
	    	app_msgnum(&nums, msgnum);
	    }
	    if (fp)
		fclose (fp);
	}
    }

    if (nums.size >= mp->numsel)
	adios (NULL, "no messages match specification");

    /*
     * So, what's happening here?
     *
     * - Re-read the folder and sequences, but with locking.
     * - Recreate the original set of selected messages from the command line
     * - Set the previous sequence
     * - Remove any messages that did NOT result in hits from the selection
     * - Add to any new sequences
     */

    if (!(mp2 = folder_read (folder, 1)))
	adios (NULL, "unable to reread folder %s", folder);

    for (msgnum = 0; msgnum < msgs.size; msgnum++)
	if (!m_convert (mp2, msgs.msgs[msgnum]))
	    done (1);
    seq_setprev (mp2);	/* set the previous-sequence */

    /*
     * Nums contains a list of messages that we did NOT match.  Remove
     * that from the selection.
     */

    for (msgnum = 0; msgnum < nums.size; msgnum++) {
    	unset_selected(mp2, nums.msgnums[msgnum]);
	mp2->numsel--;
    }

    /*
     * Add the matching messages to sequences
     */
    if (seqp > 0) {
	for (seqp = 0; seqp < svector_size (seqs); seqp++)
	    if (!seq_addsel (mp2, svector_at (seqs, seqp), publicsw, zerosw))
		done (1);
    }

    /*
     * Print total matched if not printing each matched message number.
     */
    if (!listsw) {
	printf ("%d hit%s\n", mp2->numsel, PLURALS(mp2->numsel));
    }

    svector_free (seqs);
    context_replace (pfolder, folder);	/* update current folder         */
    seq_save (mp2);			/* synchronize message sequences */
    context_save ();			/* save the context file         */
    folder_free (mp);			/* free folder/message structure */
    folder_free (mp2);
    done (0);
    return 1;
}


static void NORETURN
putzero_done (int status)
{
    if (listsw && status && !isatty (fileno (stdout)))
	puts("0");
    exit (status);
}
